### ISSUE TITLE
Mineable asteroids - actual idea

### ISSUE THREAD

**Issue by Wrzlprnft** [endless-sky/endless-sky#992]
In this [google thread](https://groups.google.com/forum/#!category-topic/endless-sky/VMt1DsWuIkw), somebody named Lyle McCracken built a mine-able asteroid, which would need just a few adjustments for the game to be actually add-able.
### Summary:
- A ship with a asteroid image without any outfits or hull, but with cargo space and exactly 1 hull
- A mining laser with .001 hull damage

=> only the mining laser can disable a asteroid

### Drawbacks:
- Still can be captured
- can transport weird stuff like food.

### Suggestion
Adding a new ship category called asteroid, and a few ifs to prevent things we don't want to happen, like
- transporting anything else than heavy metal?
- being captureable, maybe just "grey out" the capture attempt button
- having crew displayed

Along with "asteroid fleets" and a "asteroid government" this could actually be a solution to the "why can't we mine the asteroids" problem: you can mine asteroids.... but not all of them.

With some ore visible on the surface, to differ them from the asteroid tiles, this would add quite a lot to the game i think.

**Comment by Wrzlprnft** [comment]
Addition: of course, there can be a lot more done, like a "asteroid AI" or asteroid belts which are rotating and stuff, but that's too much to begin with, with the ideas displayed above it should be doable without any super-mega-effort (i hope. I am still not so far to read c++ that good i would know much about the source-code).

**Comment by Amazinite** [comment]
100% self destruction rate if you try and capture?

**Comment by Amazinite** [comment]
A work around for the strange cargo that I was planning on doing was to set the asteroids with special outfits with names like "iron" or "copper" and such. This would make it so that the asteroids hold stuff that actually makes sense, and these items could be sold at outfitters.

Drawback to this though:
There can only be as many different set ups of outfits as you have variants, which makes it predictable to know what you'll find in an asteroid (might be a downside to some but not to others).

**Comment by Wrzlprnft** [comment]
With a hidden ammunition so you cannot install them on another ship?

**Comment by Amazinite** [comment]
Yeah, that makes sense.

**Comment by LocalGod79** [comment]
i'm curious - would these "ship asteroids" be randomly everywhere, or would the player be taking jobs to mine specific asteroids?

the problem with spawning them everywhere is that at some point, the game engine would be processing "ship asteroids" all over the galaxy, burdening the cpu...

**Comment by Amazinite** [comment]
They would be fleets, so they would spawn no differently than normal AI ships.

**Comment by Wrzlprnft** [comment]
You could additionally add in some "bounty hunter" missions for asteroids, let's say laboratory needs metal x, mine it from asteroid y, or whatever. I have been thinking of asteroid fleets, too. Like "wide" up to "dense" or something like that.

**Comment by Amazinite** [comment]
If the problem is that the game continues to keep track of ships that haven't landed, even if across the galaxy, then maybe give them small amounts of thrust and turn that causes them to move towards planets and stars to "land." This could simulate asteroid impacts on planets while also cleaning the galaxy of asteroids as to not clog things up.

**Comment by LocalGod79** [comment]
the engine will only track npcs from missions - if you're having fleets as a part of the system randoms, they'll take off from the planet as well as hyperspacing into the system...

**Comment by Amazinite** [comment]
Waiting personality. 
>waiting: starts out in space in the given system, then follows you.
No engines means it can't follow you, so it just spawns in space and does nothing.

**Comment by LocalGod79** [comment]
unless it takes off from the planet...

**Comment by LocalGod79** [comment]
seems like you have this all figured out - where's the pull?

**Comment by Amazinite** [comment]
Coming soon. I haven't been home all day so I haven't had the chance to work on it.

**Comment by iMiscalculated** [comment]
This is Lyle, Thanks for taking my suggestion seriously. I was not sure it would get noticed. I will try to incorporate the things you have mentioned and post up another plugin. If you like how it ends up I would be excited to have it added.

**Comment by Amazinite** [comment]
So I've looked back over the waiting personality, and it only works on mission NPCs. So what I've done it made an invisible repeat mission that will spawn a field of asteroids (as otherwise they would just spawn out of planets) within 1 jump of the system you most recently landed on. It has a deadline of a few days as to stop a ton of asteroids from building up. Is there any way to make a mission trigger every day, just so you don't need to land to spawn an asteroid field?

**Comment by Amazinite** [comment]
Hi Lyle. I'm currently working on a version that would hopefully be able to be added. Is that fine with you, given that you seem to be wanting it in your plugin?

**Comment by iMiscalculated** [comment]
Sure, I may not have time to look at it until later this week. You probably know the core code structure better then me and would be good to get it integrated well. Good luck.

**Comment by LocalGod79** [comment]
@Amazinite unfortunately missions can only be offered on landing or boarding/assisting actions.

perhaps rather than an invisible mission, perhaps something on the job board?  tied to/unlocked by a mining corporation side plot?

that way players uninterested in mining are not having their systems loaded unnecessarily.

as i see it, people have asked for the feature so many times, some kind of implementation should be developed...

**Comment by iMiscalculated** [comment]
I was kind of thinking about something like that. 

Mixed with your idea of the Outfits the represent the ore. You mine the "Ore" for the refining companies. You take a mission that makes the asteroids appear, mine them and "Sell" the Ore as the mission reward.

Not quite the free roam mining through space, but still some variety to the activities.

**Comment by Amazinite** [comment]
I believe that both could easily be done. The current way I have it means that you'll only randomly encounter up to two asteroid fields at a time without landing, and those may even be fields that are out of the way of where you're going, so you could never see them. If someone wanted to make any real money off of mining, a job version could also be given.

"Mine this rare uranium asteroid and bring the materials back here for extra pay"
"Defend this rare asteroid from pirates until out mining ships arrive"
"Defend our mining ships from pirates"

These kinds of jobs could be how asteroids are actually acknowledged by the game, but I think that it makes sense to have random mineable asteroids every once in a while.

**Comment by Amazinite** [comment]
Quick question on fleets: The number after variant in fleets does what exactly? My guess is the frequency it appears, but is it higher or lower that makes it more common?

**Comment by LocalGod79** [comment]
a fleet variant 10 will appear 10 times as likely as a variant 1

**Comment by Amazinite** [comment]
Just thought of an issue: your escorts could blow up asteroids for no reason, even if you're trying to mine them.

**Comment by iMiscalculated** [comment]
Yes, now that you mention it. I had exactly that problem. I parked my fleet while I was testing it and didn't think about it again. Darn, not sure there is a plugin only way to get around that one.

**Comment by Amazinite** [comment]
Yeah, unless something gets coded on this to stop escorts from shooting the asteroid faction, then it's just going to be something that people should worry about.

**Comment by LocalGod79** [comment]
this is where the job comes in - the player can have a mining fleet without normal weapons, and the other fleet, which gets parked for the job.

#645 might also solve that problem.

**Comment by Amazinite** [comment]
Yeah, I have the random asteroid fields done and I'm working on a few jobs.

**Comment by Amazinite** [comment]
[asteriods.txt](https://github.com/endless-sky/endless-sky/files/233781/asteriods.txt)
I haven't tested anything yet, and I don't have any more time today to work. I'll test it tomorrow and make an actual pull for it if everything is good.

**Comment by Amazinite** [comment]
All it really needs are images for the materials and for the mining laser.

 *cough*

**Comment by LocalGod79** [comment]
you want to take this one @Wrzlprnft ?  my hands are already full with he Boxwing...

**Comment by Wrzlprnft** [comment]
How about splitting it up? i am working on a set of outfits at the moment, so one more or less, like a mining laser and/or a mining laser turret won't matter that much, and i can create a beam and maybe a hit effect, too (we want to match it with the laser, won't we), but i would be glad if you could do the materials.

**Comment by Amazinite** [comment]
[asteriods.txt](https://github.com/endless-sky/endless-sky/files/235568/asteriods.txt)
I've fixed a ton of things up and balanced out the missions, but I can't seem to get the escort missions to appear. I'm not getting any errors, so can anyone see what is wrong?

**Comment by LocalGod79** [comment]
combat rating less than 300?

**Comment by Amazinite** [comment]
@LocalGod79 Thanks, that worked. They are appearing now. Only things left to do are to make fine tune balances and get the images.

**Comment by Amazinite** [comment]
These are the descriptions I've come up with for the mining lasers. They might help to give you ideas for when you're making images. 

Mining Laser:
During the early 22nd century, asteroid mining was a common practice among humanity. Resources were plentiful and the labor was cheap. But during the vast expansion of humanity, asteroid mining was left behind for better mining operation on newly found planets. 
The Syndicate saw this old practice as a perfect way to open up a new market. Thus the mining laser was created, a low power laser intended for the use of bursting away the outer layers of asteroids to reveal the precious minerals beneath the crust without destroying the asteroid itself. 

Advanced Mining Laser:
As the Syndicate started to mine away the asteroid fields in some systems, they began running into bigger and tougher asteroids. It began taking hours for a single mining laser to reveal enough minerals on an asteroid to make profit. In response, the Syndicate developed the advanced mining laser, which outputs more energy than the normal mining laser and is able to reveal ten times the material in the same amount of time.  

Mining Turret:
When mining ships began to grow bigger and bigger in order to hold more ore, their speeds began to get slower and slower. This made mining asteroids a real pain due to the long amounts of time spent turning the ship to keep in aligned with the asteroids. As a result, miners began strapping their lasers to old blaster turret platforms to get more control out of their lasers. 

Advanced Mining Turret:
Given that advanced mining lasers are much more expensive than normal mining lasers, no miners were willing to possibly waste their advanced mining lasers creating turrets. In response, Syndicate decided to create their own turrets fit with dual advanced mining lasers for maximum production.

**Comment by iMiscalculated** [comment]
I was looking at the code and I think it would be pretty easy to set up a conditional that would grey out the capture button. I just need to down load the code and figure out how to compile it to try it out.

**Comment by Amazinite** [comment]
Mineral descriptions. I'm not sure if some of these might break lore, namely the titanium one. I have no idea that the hull of ships are made out of. 

Copper:
Copper is a common metal found in almost every asteroid in small amounts. Even so, it still goes for a fair price given its high demand for use in electrical wiring. 

Iron:
Iron is not as common as copper in asteroids, but is just as highly demanded. Iron is used in the creation of steel, an important material for building planetary infrastructure and many of the parts and outfits for ships. 

Silver:
Due to silver's spectacular properties, it is used for a wide number of things, such as more expensive electrical wiring in larger ships, as a key material in photovoltaic panels, and perhaps most importantly, as a cheap way of purifying water while in the depths of space. 

Titanium:
Titanium is one of the main materials use in creating the hulls of ships. Flexible, durable, and strong, titanium is sure to go for a high price anywhere on the market. 

Gold:
Although gold has lost some value over the years due to planetary and asteroid mining making it much more common than in the early days of humanity's expansion, it is still a highly sought after mineral due to its beauty and demand in jewelry. 

Gemstones:
Gemstones are little more than a vanity item, having little use in the creation of ship parts. Even so, there are a number of citizens in the Paradise Worlds willing to pay big money for some genuine space gemstones. 

Uranium
Uranium is an important element used in the creation of some reactors. Much of the uranium mines from asteroids and sold on the market is usually bought up by the Syndicate, but why that is is not exactly known. 

Unknown Material
This unknown material was mined from an asteroid near Syndicate space. It gives off a faint humming sound and glows a low red color when placed in the dark. You aren't exactly sure what it is, but you aren't willing to mess around it to find it. Better return this to the scientists who sent you out to mine it.

**Comment by LocalGod79** [comment]
i think that many of these images, you may be able to source yourself (a pile of creek bed gold nuggets could be coloured for most of the ores), if you're willing to go digging through public domain sources, and note the credits.

**Comment by Amazinite** [comment]
Okay, I'll try and get images for the materials later today. I haven't messed with blender yet and there's no way my gimp images are going to fit in to the game, so could you please do that? I believe Wrzl may have started working on something though.

**Comment by Wrzlprnft** [comment]
I have a pretty solid idea about the laser, and i can make the other stuff too, but for that i may need the weekend.

**Comment by endless-sky** [comment]
I'm worried that if this were added to the base game as-is, pretty much everyone would be disappointed - the players who want "real" mining would find the disabling and boarding mechanism way too contrived, and the players who don't expect asteroids to be destructible would be confused about why some of them suddenly are. I'd rather spend time to plan out and implement a high-quality mining implementation with good mechanics, then rush to add something that feels like a hack.

That said, I'm sure people would appreciate this as a plugin, and any work you did in the plugin (e.g. making sprites and descriptions for the various minerals) could be ported to the main game once a "final" mining implementation is in place (which, conceivably, could be 0.9.1; see below). In particular, you could probably find good outfit images with CC-BY-SA licenses on Wikipedia, e.g.:

https://commons.wikimedia.org/wiki/Commons:Robert_Lavinsky
https://commons.wikimedia.org/wiki/Category:Valued_images_of_minerals

In terms of a "final" mining implementation, here's my proposal:

* Every time you enter a system, the game will spawn a small number of mineable asteroids that are visually distinct from the ordinary, indestructible ones.
* Those asteroids travel in circular orbits rather than along linear paths, so that they won’t all drift so far away from the system center that you can’t find them.
* Each system will define what mineable asteroids it contains, allowing the more valuable ones to be rare and to be placed in dangerous or remote locations.
* Mineable asteroids can be destroyed by any weapon that does hull damage.
* When they are destroyed, they create small “flotsam” objects that any ship with free cargo space can pick up just by colliding with them.
* Any “flotsam” you pick up is converted into special outfits in your cargo hold: “iron ore,” etc. The same mechanism could also be used to let you pick up cargo dumped by exploding ships.
* Those outfits are marked as something you cannot install, only sell.
* The trading panel will list those outfits in cargo as "harvested materials" rather than "plundered outfits."
* Each type of mineable material will be provided by an asteroid with a slightly different appearance, so you know what you’re going to get before you blow it up.
* There will also be AI ships that mine asteroids, both to show the player it can be done and to add realism.

**Comment by endless-sky** [comment]
For a start, see commit e80796d, which will make outfits with the "installable" attribute set to -1 display as "harvested materials" in the trading panel (in addition to being incapable of being installed in a ship).

**Comment by Amazinite** [comment]
I'm glad you're on board with the idea.
I completely agree that what I have in place doesn't do full justice to what some people want. It was more of an experiment to see what could be done, but I don't exactly have the tools with the current game engine to do everything right now. Much of what you listed is exactly what I would want to see in the main game, so it's great that it's pretty much roadmapped now to eventually happen.

**Comment by endless-sky** [comment]
Here's an example of using a mineral image from Wikipedia. I had to boost the contrast and paint some shadows on the bottom of it to make it fit in with the other outfit graphics:

![pyrite](https://cloud.githubusercontent.com/assets/11450262/14874454/40bd8474-0cce-11e6-95d3-43a20c1fd08f.png)

**Comment by nylki** [comment]
> Those asteroids travel in circular orbits rather than along linear paths, so that they won’t all drift so far away from the system center that you can’t find them.

After all, my gravity fork may come in handy ;P ( https://github.com/endless-sky/endless-sky/issues/912 )

**Comment by Wrzlprnft** [comment]
Since the actual idea this issue was about is, due to @endless-sky 's awesomeness no longer relevant, ill close this.
